### PR TITLE
[TableGen] Bug fix for tied optional operands resolution

### PR DESCRIPTION
**Pull request by AlfieRichardsArm**
This fixes tied operand resolution when there are optional operands before it.

### CODE REVIEW

**Comment by AlfieRichardsArm** [comment]
@s-barannikov Pulled this bug fix out as a separate patch

File: llvm/utils/TableGen/AsmMatcherEmitter.cpp [hunk 1]
```diff
@@ -1986,9 +1986,9 @@ emitConvertFuncs(CodeGenTarget &Target, StringRef ClassName,
   }
   CvtOS << "  assert(Kind < CVT_NUM_SIGNATURES && \"Invalid signature!\");\n";
   CvtOS << "  const uint8_t *Converter = ConversionTable[Kind];\n";
-  CvtOS << "  unsigned OpIdx;\n";
   CvtOS << "  Inst.setOpcode(Opcode);\n";
   CvtOS << "  for (const uint8_t *p = Converter; *p; p += 2) {\n";
+  CvtOS << "    unsigned OpIdx;\n";
```
Comment (s-barannikov) [inline llvm/utils/TableGen/AsmMatcherEmitter.cpp:1991]: Since you're moving the declaration, you can also fuse fuse it with the definition (`unsigned OpIdx = ...`).

File: llvm/utils/TableGen/AsmMatcherEmitter.cpp [hunk 2]
```diff
@@ -1999,16 +1999,17 @@ emitConvertFuncs(CodeGenTarget &Target, StringRef ClassName,
   }
   CvtOS << "    switch (*p) {\n";
   CvtOS << "    default: llvm_unreachable(\"invalid conversion entry!\");\n";
-  CvtOS << "    case CVT_Reg:\n";
+  CvtOS << "    case CVT_Reg:{\n";
```
Comment (s-barannikov) [inline llvm/utils/TableGen/AsmMatcherEmitter.cpp:2002]: Unnecessary change.

**Review by s-barannikov — APPROVED** [review]
Could you add a FIXME near OpIdx calculation that it is not an operand index in CVT_Tied case?

File: llvm/utils/TableGen/AsmMatcherEmitter.cpp [hunk 3]
```diff
@@ -1992,6 +1992,7 @@ emitConvertFuncs(CodeGenTarget &Target, StringRef ClassName,
     // When optional operands are involved, formal and actual operand indices
     // may differ. Map the former to the latter by subtracting the number of
     // absent optional operands.
+    // FIXME: This is not an operand index in the CVT_Tied case
```
Comment (s-barannikov) [inline llvm/utils/TableGen/AsmMatcherEmitter.cpp:1995]: Nice, thanks.

File: llvm/utils/TableGen/AsmMatcherEmitter.cpp (hunk 1 above)
Comment (AlfieRichardsArm) [reply llvm/utils/TableGen/AsmMatcherEmitter.cpp]: Done
